engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java

```diff
@@ -262,47 +276,144 @@ public Class<?> getVariableType(final String var) {
     }
 
 
-    @Override
     public TableDefinition getTableDefinition(final String var) {
         Object o = getVariable(var, null);
         return o instanceof Table ? ((Table) o).getDefinition() : null;
     }
 
+
+    /**
+     * Retrieve a variable from the script session's bindings.
+     * <p/>
+     * Please use {@link ScriptSession#getVariable(String, Object)} if you expect the variable may not exist.
+     *
+     * @param name the variable to retrieve
+     * @return the variable
+     * @throws QueryScope.MissingVariableException if the variable does not exist
+     */
+    @NotNull
+    protected abstract Object getVariable(String name) throws QueryScope.MissingVariableException;
+
+    /**
+     * Retrieve a variable from the script session's bindings. If the variable is not present, return defaultValue.
+     * <p>
+     * If the variable is present, but is not of type (T), a ClassCastException may result.
+     *
+     * @param name the variable to retrieve
+     * @param defaultValue the value to use when no value is present in the session's scope
+     * @param <T> the type of the variable
+     * @return the value of the variable, or defaultValue if not present
+     */
+    protected abstract <T> T getVariable(String name, T defaultValue);
+
+    /**
+     * Retrieves all of the variables present in the session's scope (e.g., Groovy binding, Python globals()).
+     *
+     * @return an unmodifiable map with variable names as the keys, and the Objects as the result
+     */
+    protected abstract Map<String, Object> getVariables();
+
+    /**
+     * Retrieves all of the variable names present in the session's scope
+     *
+     * @return an unmodifiable set of variable names
+     */
+    protected abstract Set<String> getVariableNames();
+
+    /**
+     * Check if the scope has the given variable name
+     *
+     * @param name the variable name
+     * @return True iff the scope has the given variable name
+     */
+    protected abstract boolean hasVariableName(String name);
+
+    /**
+     * Inserts a value into the script's scope.
+     *
+     * @param name the variable name to set
+     * @param value the new value of the variable
+     */
+    protected abstract void setVariable(String name, @Nullable Object value);
+
     @Override
     public VariableProvider getVariableProvider() {
-        return this;
-    }
+        return new VariableProvider() {
```

Comment:
Arguably this interface and impl should go away, and QueryScope should be used by autocomplete instead. At this point it still exists to allow direct access to script session variables instead of the "more correct" way of going through the QueryScope.

Note that the abstract methods above are to be implemented by each language, but the locking/etc is handled by the VariableProvider/QueryScope, so those are the public interfaces to get access to data.

Reply:
I think it makes sense to do it now, but I wanted to raise the possibility of keeping separate APIs, with "this is for query scope, which comes from the exec context, and might not be related to a script" vs "this is a specific way to interact with variables in the script session".

Another option (that we're both leaning towards) is to let ScriptSession have a public `getQueryScope()`, so that code that has a script session can be more generic...